Libraries/LibGfx/Quad.h

```diff
@@ -29,20 +28,45 @@ class Quad {
 
     Rect<T> bounding_rect() const
     {
-        auto top = min(min(m_p1.y(), m_p2.y()), min(m_p3.y(), m_p4.y()));
-        auto right = max(max(m_p1.x(), m_p2.x()), max(m_p3.x(), m_p4.x()));
-        auto bottom = max(max(m_p1.y(), m_p2.y()), max(m_p3.y(), m_p4.y()));
-        auto left = min(min(m_p1.x(), m_p2.x()), min(m_p3.x(), m_p4.x()));
-        return { left, top, right - left, bottom - top };
+        T left = min(min(m_p1.x(), m_p2.x()), min(m_p3.x(), m_p4.x()));
+        T right = max(max(m_p1.x(), m_p2.x()), max(m_p3.x(), m_p4.x()));
+        T width = right - left;
+
+        T top = min(min(m_p1.y(), m_p2.y()), min(m_p3.y(), m_p4.y()));
+        T bottom = max(max(m_p1.y(), m_p2.y()), max(m_p3.y(), m_p4.y()));
+        T height = bottom - top;
+
+        return { left, top, width, height };
     }
 
     bool contains(Point<T> point) const
     {
-        // FIXME: There's probably a smarter way to do this.
-        return Triangle(m_p1, m_p2, m_p3).contains(point)
-            || Triangle(m_p1, m_p3, m_p4).contains(point)
-            || Triangle(m_p2, m_p4, m_p1).contains(point)
-            || Triangle(m_p2, m_p4, m_p3).contains(point);
+        // Even-Odd algorithm: https://www.geeksforgeeks.org/even-odd-method-winding-number-method-inside-outside-test-of-a-polygon/
+        //
+        // 1. "Constructing a line segment between the point (P) to be examined and a known point outside the polygon"
+        //      - We're using horizontal line from (point.x, point.y) to (bounding_rect().left + bounding_rect().width + 1, point.y)
+        //        (i.e. just +1 to right of furthest-right point in quad)
+        //
+        // 2. "The number of times the line segment intersects the polygon boundary is then counted."
+        //      - We count the line's intersections with the quad by checking each quad edge for intersection (1-2, 2-3, 3-4, 4-1)
+        //
+        // 3. "The point (P) is an internal point if the number of polygon edges intersected by this line is odd;
+        //    otherwise, the point is an external point."
+
+        u8 n_intersections_with_quad = 0;
+        Point<T> const* quad_points[5] = { &m_p1, &m_p2, &m_p3, &m_p4, &m_p1 };
```

Comment:
Could probably be a simple `AK::Array` as well? Gets some bounds checking in there, although that's hard to get wrong in the loop.